File: GeeksForGeeks/src/main/java/_60Days/SearchSort/BubbleSort.java
package _60Days.SearchSort;

public class BubbleSort {

	public static void main(String[] args) {
		int[] aR = { 1, 2, 7, 3, 1, 5, 7, 8, 3, 5, 2, 6 };

		bubbleSortOptamized(aR);

		bubbleSortNaieve(aR);

		for (int x : aR) {
			System.out.print(x + " ");
		}
	}

	// O(N^2) - Worst
	// O(N) Best
	private static void bubbleSortOptamized(int[] nums) {
		for (int i = 0; i < nums.length; i++) {
			boolean hasChange = false;
			for (int j = 0; j < nums.length - 1; j++) {
				if (nums[i] < nums[j]) {
					hasChange = true;
					int Temp = nums[i];
					nums[i] = nums[j];
					nums[j] = Temp;
				}
			}
			if (!hasChange)
				break;
		}

	}

	// O(N^2)
	private static void bubbleSortNaieve(int[] nums) {
		for (int i = 0; i < nums.length; i++) {
			for (int j = 0; j < nums.length - 1; j++) {
				if (nums[i] < nums[j]) {
					int Temp = nums[i];
					nums[i] = nums[j];
					nums[j] = Temp;
				}
			}
		}
	}

}

File: Algoexpert/src/main/kotlin/easy/BubbleSort.kt
package easy

class BubbleSort {

    fun bubbleSort(array: MutableList<Int>): List<Int> {
        var i=0
        while(i<array.size){
            var j = 0
            while(j < array.size-1){
                if(array[j] > array[j+1]){
                    var temp = array[j]
                    array[j] = array[j+1]
                    array[j+1] = temp
                }
                j++
            }
            i++
        }
        return array
    }

    fun bubbleSort_(array: MutableList<Int>): List<Int> {
        var i=0
        while(i<array.size){
            var j = 0
            var isSorted = true
            while(j < array.size -1){
                if(array[j] > array[j+1]){
                    var temp = array[j]
                    array[j] = array[j+1]
                    array[j+1] = temp
                    isSorted = false
                }
                j++
            }
            if(isSorted) return array
            i++
        }
        return array
    }
}